src/chainbin/nnet3-chain-get-egs-post.cc

```diff
@@ -0,0 +1,397 @@
+  typedef fst::StdArc ArcOut;
+  typedef ArcIn::StateId StateId;
+  ofst->DeleteStates();
+  // The states will be numbered exactly the same as the original FST.
+  // Add the states to the new FST.
+  StateId num_states = ifst.NumStates();
+  for (StateId s = 0; s < num_states; s++) {
+    StateId news = ofst->AddState();
+    assert(news == s);
+  }
+  ofst->SetStart(ifst.Start());
+  for (StateId s = 0; s < num_states; s++) {
+    LatticeWeight final_iweight = ifst.Final(s);
+    if (final_iweight != LatticeWeight::Zero()) {
+      fst::TropicalWeight final_oweight;
+      ConvertLatticeWeight(final_iweight, &final_oweight);
+      ofst->SetFinal(s, final_oweight);
+    }
+    for (fst::ArcIterator<Lattice> iter(ifst, s);
+         !iter.Done();
+         iter.Next()) {
+      ArcIn arc = iter.Value();
+      KALDI_PARANOID_ASSERT(arc.weight != LatticeWeight::Zero());
+      ArcOut oarc;
+      ConvertLatticeWeight(arc.weight, &oarc.weight);
+      oarc.ilabel = tmodel.TransitionIdToPdf(arc.ilabel) + 1;
+      oarc.olabel = tmodel.TransitionIdToPdf(arc.ilabel) + 1;
+      oarc.nextstate = arc.nextstate;
+      ofst->AddArc(s, oarc);
+    }
+  }
+}
+
+
+/**
+   This function does all the processing for one utterance, and outputs the
+   supervision objects to 'example_writer'.  Note: if normalization_fst is the
+   empty FST (with no states), it skips the final stage of egs preparation and
+   you should do it later with nnet3-chain-normalize-egs.
+*/
+
+static bool ProcessFile(const fst::StdVectorFst &normalization_fst,
+                        const GeneralMatrix &feats,
+                        const MatrixBase<BaseFloat> *ivector_feats,
+                        int32 ivector_period,
+                        const Lattice &lat,
+                        int32 num_output_frames,
+                        const std::string &utt_id,
+                        bool compress,
+                        int32 num_pdfs,
+                        TransitionModel &tmodel,
+                        UtteranceSplitter *utt_splitter,
+                        NnetExampleWriter *example_writer) {
+  //KALDI_ASSERT(supervision.num_sequences == 1);
+  int32 num_input_frames = feats.NumRows();
+
+  if (!utt_splitter->LengthsMatch(utt_id, num_input_frames, num_output_frames))
+    return false;  // LengthsMatch() will have printed a warning.
+
+  std::vector<ChunkTimeInfo> chunks;
+
+  utt_splitter->GetChunksForUtterance(num_input_frames, &chunks);
+
+  if (chunks.empty()) {
+    KALDI_WARN << "Not producing egs for utterance " << utt_id
+               << " because it is too short: "
+               << num_input_frames << " frames.";
+    return false;
+  }
+
+  int32 frame_subsampling_factor = utt_splitter->Config().frame_subsampling_factor;
+
+  fst::StdVectorFst sup_fst,
+    scaled_normalization_fst(normalization_fst);
+  ConvertLatticeToPdfLabels(tmodel, lat, &sup_fst);
+  ScaleFst(0.5, &scaled_normalization_fst); // Scale lattice to have weights similar
```

Comment:
Instead to doing this scaling for each utterance, you can do it immediately after reading the fst.